scio-core/src/main/scala/com/spotify/scio/values/PairSCollectionFunctions.scala

```diff
@@ -488,6 +488,55 @@ class PairSCollectionFunctions[K, V](val self: SCollection[(K, V)]) {
     }
   }
 
+  /**
+   * Return an SCollection with the pairs from `this` whose keys are in `that`
+   * when the cardinality of `this` >> `that`, but neither can fit in memory
+   * (see [[PairHashSCollectionFunctions.hashIntersectByKey]]).
+   * @group per key
+   */
+  def sparseIntersectByKey(that: SCollection[K], thatNumKeys: Int, fpProb: Double = 0.1)(
+    implicit koder: Coder[K],
+    voder: Coder[V],
+    hash: Hash128[K]): SCollection[(K, V)] = {
+    val bfSettings =
+      PairSCollectionFunctions.optimalBFSettings(thatNumKeys, fpProb)
+    if (bfSettings.numBFs == 1) {
+      sparseIntersectByKeyImpl(that, thatNumKeys.toInt, fpProb)
+    } else {
+      val n = bfSettings.numBFs
+      val thisParts = self.partition(n, _._1.hashCode() % n)
+      val thatParts = that.partition(n, _.hashCode() % n)
+      val joined = (thisParts zip thatParts).map {
+        case (lhs, rhs) =>
+          lhs.sparseIntersectByKeyImpl(rhs, bfSettings.capacity, fpProb)
+      }
+      SCollection.unionAll(joined)
+    }
+  }
+
+  protected def sparseIntersectByKeyImpl(that: SCollection[K],
+                                         thatNumKeys: Int,
+                                         fpProb: Double = 0.1)(
+    implicit koder: Coder[K],
+    voder: Coder[V],
+    hash: Hash128[K]): SCollection[(K, V)] = {
+    val width = BloomFilter.optimalWidth(thatNumKeys, fpProb).get
+    val numHashes = BloomFilter.optimalNumHashes(thatNumKeys, width)
+    val rhsBf = that
+      .aggregate(BloomFilterAggregator[K](numHashes, width))
+      .asIterableSideInput
```

Comment:
Any reason not using singleton side input here?

Reply:
as a singleton, in the case of an empty `that` it'll fail with `Empty PCollection accessed as a singleton view` when we try to access `rhsBf` through side input context. I based it off of the SparseJoin implementations, which use an `IterableSideInput` in the same way. wdyt?

Reply:
Actually there's a `def asSingletonSideInput(defaultValue: T): SideInput[T]` with default now maybe use that for both?

Reply:
sounds good! added a unit test for that case in SparseOuterJoins too. Also added an `exact` boolean to the sparse intersection.